docs/decisions/0011-depending-on-multiple-event-bus-implementations.rst

```diff
@@ -0,0 +1,80 @@
+- This approach does not preclude the use of other implementations that are not part of the Open edX project. If a community member creates their own implementation and it is not offered to or accepted into the Open edX project, it can still be installed out-of-tree by that deployer.
+
+There are also some downsides:
+
+- The dependency tree is larger, and is more or less guaranteed to include at least one top-level dependency that each deployer is not using. Not only does each dependency take up bandwidth to download, space on disk, and time spent in upgrading packages, but it can also bring licensing and architecture incompatibilities.
+- More concretely, event-bus-kafka depends on confluent-kafka, but that package does not provide binaries compatible with Linux on Apple Silicon. This means that developers on Apple M1 laptops cannot install this package in devstack at all. So as a knock-on effect, event-bus-kafka has had to keep confluent-kafka out of its own in-tree dependencies; deployers wishing to use it have to install and version that package out-of-tree anyhow. While this may eventually be resolved, this does obviate a number of the stated benefits for development and version management for the Kafka users.
+
+We are not necessarily *satisfied* with this approach, but regard it as suitable for at least one release.
+
+Rejected Alternatives
+*********************
+
+There are several alternative approaches we have considered:
+
+- Have the implementation be part of the openedx-events package
+- Just have one implementation
+- Require all implementations to be installed out-of-tree
+- Depend on one default implementation in-tree, and require that others be installed
+
+Implement inside openedx-events
+===============================
+
+We briefly discussed the idea of including all of the implementations *in* the openedx-events package, possibly as extras (e.g. ``openedx-events[kafka]``). However, this would have complicated the development of openedx-events; it would have become impossible to e.g. pin the version of the Redis implementation in a dependent IDA while continuing to receive updates to the Kafka implementation. (The use of extras also would not have helped our dependencies issue; we'd then need to decide which extras to depend upon from IDAs' base requirements files.)
+
+Single implementation
+=====================
+
+It's very unlikely that one implementation would satisfy everyone. Any event store, queue, or message bus technology that is used is additional software that deployers must learn to install, manage, configure, and operate. The exceptions would be MySQL and Redis, which are already used in Open edX. However, neither of these was used for the first implementation. Therefore, even if Redis turns out to work very well, we'll still need the extension point.
```

Comment:
According to [the ADR](https://github.com/openedx/event-bus-kafka/blob/main/docs/decisions/0002-kafka-based-event-bus.rst), the cons for Redis were that it can "Can lose acked data, even if RAM backed up with an append-only file (AOF)." (unfortunately no details were provided, so I can't tell if this claim is still relevant), and that it "Requires homegrown schema management." (which I think we are doing anyways?)

Reply:
Hah! Of course I should have checked previous ADRs, thank you. :-)

Reply:
A more current ADR also exists for why redis (vs why not redis): https://github.com/openedx/event-bus-redis/blob/main/docs/decisions/0002-redis-streams-based-event-bus.rst